mantidproject/mantid
# ISIS Powder improvements for using non-standard grouping files

HRPD have a number of non-standard grouping files that they like to use. Doing this makes it reasonably easy to run into errors where there is a mismatch between the numbers of banks in the various inputs. This PR improves the readability of these errors and also enables dividing by a vanadium spline with only one bank.

**To test:**

As with all `isis_powder` PRs, if you've not set up your directories and config files, give me a poke and I'll give you a hand. Sample data is [here](https://github.com/mantidproject/mantid/files/1501831/HRPD.zip).

Create your HRPD instrument using the attached config file (you'll need to set the paths in it to suit your setup)

```
from isis_powder import HRPD
hrpd = HRPD(config_file=r"path/to/hrpd_config.yaml")
```
First, create your vanadium and do a focus normally. This should give you no errors, the output workspaces should have nice sharp peaks
```
hrpd.create_vanadium(first_cycle_run_no=68236, window="10-110")
hrpd.focus(run_number=68245, window="10-110")
```
Put `hrpd_new_072_01_corr_simm_left3_test.cal` (attached) in your calibration directory, and then try and do a focus using it:
```
hrpd.focus(run_number=68245, window="10-110",
           grouping_file_name="hrpd_new_072_01_corr_simm_left3_test.cal")
```
You should get:
```
RuntimeError: Mismatch between number of banks in vanadium and number of banks in workspace to focus
There are 3 banks for vanadium but 1 for the run
```
Now try creating your vanadium with the new grouping file and focusing with the correct number of cropping values and bin widths:
```
hrpd.create_vanadium(first_cycle_run_no=68236, window="10-110",
                     grouping_file_name="hrpd_new_072_01_corr_simm_left3_test.cal")
hrpd.focus(run_number=68245, window="10-110", focused_bin_widths=[-0.0005],
           focused_cropping_values=[(2e4, 1e5)], vanadium_normalisation=True)
```
This should not throw any errors, and you should again get nicely defined peaks. Check that the workspace `68245.raw-Results-TOF-Grp` is cropped according to the parameters in the script.

Fixes #21259 


**Release Notes** 
[Here](https://github.com/mantidproject/mantid/commit/7c5cec855bbf9d5aa71ea32e02da31a22ac64f34)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
